build: add linker optimisation flags to gitian & guix (Linux)

fanquake opened the pull request:
This PR adds `-Wl,O2` to our gitian and guix LDFLAGS. This makes the linker perform certain optimisations (and is different from LTO).

Any -O argument will enable optimizations in GNU ld. We can use -O2 here, as this matches our compile flags. Note that this would also enable additional optimizations if using the lld or gold linkers, when compared to -O0.

A nice writeup + diagrams of some of these optimizations is  available here: http://lwn.net/Articles/192624/.

#### master
```bash
# bitcoind
Histogram for `.gnu.hash' bucket list length (total of 3 buckets)
 Length  Number     % of total  Coverage
      0  1          ( 33.3%)       0.0%
      1  0          (  0.0%)       0.0%
      2  1          ( 33.3%)      40.0%
      3  1          ( 33.3%)     100.0%
```
```bash
# bitcoin-qt 
Histogram for `.gnu.hash' bucket list length (total of 3 buckets)
 Length  Number     % of total  Coverage
      0  0          (  0.0%)       0.0%
      1  1          ( 33.3%)      10.0%
      2  0          (  0.0%)      10.0%
      3  0          (  0.0%)      10.0%
      4  1          ( 33.3%)      50.0%
      5  1          ( 33.3%)     100.0%
```

#### this PR:
```bash
# bitcoind
Histogram for `.gnu.hash' bucket list length (total of 8 buckets)
 Length  Number     % of total  Coverage
      0  3          ( 37.5%)       0.0%
      1  5          ( 62.5%)     100.0%
```
```bash
# bitcoin-qt
Histogram for `.gnu.hash' bucket list length (total of 19 buckets)
 Length  Number     % of total  Coverage
      0  9          ( 47.4%)       0.0%
      1  10         ( 52.6%)     100.0%
```

#### GNU ld -O

> If level is a numeric values greater than zero ld optimizes the output. This might take significantly longer and therefore probably should only be enabled for the final binary. At the moment this option only affects ELF shared library generation. Future releases of the linker may make more use of this option. Also currently there is no difference in the linker’s behaviour for different non-zero values of this option. Again this may change with future releases.

#### lld -O

> Optimize output file size

fanquake committed:
build: add linker optimization flags to gitian descriptors

Any -O argument will enable optimizations in GNU ld. We can use -O2
here, as this matches our compile flags. Note that this would also
enable additional optimizations if using the lld or gold linkers,
when compared to -O0.

MarcoFalke commented:
How is this different from #16791?

fanquake commented:
> How is this different from #16791?

@MarcoFalke Despite the similar names, this is not the same thing as LTO. This will optimize the symbol hash tables in binaries, whereas LTO does intermodular optimizations. More info on either: [Optimizing Linker Load Times](https://lwn.net/Articles/192624/) and [LTO](https://llvm.org/docs/LinkTimeOptimization.html).

dongcarl commented:
Concept ACK, good to know that it's different from LTO.

Need to look into it more carefully.

dongcarl commented:
Concept ACK, although would like to see this in Gitian/release builds instead of default configure. At some point we should probably add a doc to the repo of recommended configure flags and what they do.

-----

It seems to me that, when supplied with `-O2`:

- `ld` optimizes the hash chain size for the symbol hash table, which may lead to larger outputs
- `ld.gold` optimizes the output file size

Perhaps that accounts for the performance differences you're seeing?

fanquake commented:
> Concept ACK, although would like to see this in Gitian/release builds instead of default configure. 

I agree, and I've change this to be a configure flag, `--enable-linker-optimizations` (off by default). Then we can pass it into gitian builds in future if desired. Going to follow up with some more info shortly.

practicalswift commented:
Concept ACK

jgarzik commented:
Passing optimization flags to the link stage has been the default for decades in autotools projects (autoconf/automake/libtool).   Thus,

1) concept ACK
2) It should be enabled by default.

dongcarl commented:
From #bitcoin-builds meeting notes today: We should switch to just adding the -Wl,-O2 to LDFLAGs in Gitian/Guix, and benchmark to see the differences.

fanquake committed:
build: add linker optimization flags to guix

Any -O argument will enable optimizations in GNU ld. We can use -O2
here, as this matches our compile flags. Note that this would also
enable additional optimizations if using the lld or gold linkers,
when compared to -O0.

fanquake commented:
I've updated this PR as discussed. Add the flags to the guix and gitian descriptors. Also adding additional information to the PR description.

dongcarl commented:
ACK f2b5b0a3b48f9241971c14238040048f8b630811
Thanks for the histogram!

laanwj commented:
ACK f2b5b0a3b48f9241971c14238040048f8b630811
